File: ex1/gradientDescent.m
function [theta]=gradientDescent(X,y,alpha,iterations,theta)
%%梯度下降法求J最小的参数

x=X(:,2);
theta(1)=0;
m = length(y); 
for i =1:iterations
    theta0=theta(1);
    theta1=theta(2);
    Hy=X*theta;
    theta0=theta0-alpha/m*sum(Hy-y);
    theta1=theta1-alpha/m*sum((Hy-y).*x);
    theta=[theta0;theta1];
end

end

File: ex3/Multi-class Classification_python/gradient_descent.py
import numpy as np
import sigmoid as S

def gradient_descent(theta,X,y,lam):
    """
    该函数只需要求得偏导数(正则化之后)
    """
    theta = theta.T
    g = S.sigmoid(np.dot(X,theta))
    grad = (np.dot(X.T,g - y)/len(X) ).flatten()  #J的偏导数矩阵gard

    theta_Reg = theta[1:]             # 不惩罚第一项
    grad_Reg = np.zeros(theta.shape)
    grad_Reg[0] = grad[0]
    # 对除theta0外部分参数进行正则化处理
    grad_Reg[1:] = grad[1:] + lam * theta_Reg / len(X)

    return grad_Reg